### DISCUSSION TITLE
Is it possible to call async function more than once at certain time interval?

### DISCUSSION THREAD

**Discussion by ujur007** [hbldh/bleak#1091] [Q&A]
* bleak version:Latest
* Python version:3.8.10
* Operating System: Linux Mint-64bit
* BlueZ version (`bluetoothctl -v`) in case of Linux: 5.56

### Description
I realized that calling the async function more than once or creating two saparate async function gives a runtime error "event loop is already closed". While searching for solution I found the suggestion to move everything with await in one function. While I may want to create another async function or call the same function more than once to see if the device is discover or new devices are available.

**Comment by dlech** [comment]
There should only ever be one run loop run in your application. Please see the example programs for how to write a program that uses a single call to `asyncio.run()` to do this.

You can use `asyncio.create_task()`, `asyncio.gather()`, `asyncio.wait()`, etc. to run multiple tasks in parallel.

**Comment by ujur007** [comment]
Thanks. Do you refer to any specific example in this case? because, I think that most of the examples are pretty basics.
What I want to see is how to wait for some unknown time.

I am actually working on creating an application. The example scenario would be to let's say first check the available devices, then to connect to one of then and then send some values for the characteristics. The delay in this case would be user input based.

**Reply by dlech** [reply]
> What I want to see is how to wait for some unknown time.

This is generally done with `asyncio.Event` if you don't need to pass any information from the trigger to the waiter or `asyncio.Future` if you do need to pass a value.

**Reply by ujur007** [reply]
I tried with asyncio.Event, await, create_task etc but as I am using kivy buttons where a callback must be defined with 'def' syntax only this seems not possible to me. Cause when the awaitables are running everything else just stops.

**Reply by dlech** [reply]
Can you share a simple reproducible example code?

**Reply by ujur007** [reply]
Yes, sure. The following is a part of my code where I have tasks running in event loop. The scanview_task does not finish and showrgbmatrix does not run at all. I have to define btn_press methods with def syntax and then wait in other task using either while(variable=True) loop or other methods to wait until task can run till end.

   
   ```
def __init__(self, **kwargs):
        self.disconnect = False
        super(DeviceList,self).__init__(**kwargs)
        self.cols = 2
        asyncio.run(self.scandevices())

    async def scandevices(self):
        self.event = asyncio.Event()
        scan_task = asyncio.create_task(self.scanview_task(self.event))
        await scan_task
        
        self.event2 = asyncio.Event()
        wait_task = asyncio.create_task(self.waiting(self.event2))
        await wait_task
        
    async def scanview_task(self, event):        
        self.searchdevices = await BleakScanner.discover(5)
        for i in self.searchdevices:
            self.connectbutton = Button(text = 'Connect', size_hint =(0.2,0.4), pos_hint = {'center_x' : 0.5, 'center_y':0.5})
            self.address = i.address
            client = BleakClient(self.address)            
            self.connectbutton.bind(on_press = partial(self.showrgbmatrix, client))
            self.label = Label(text = str(i.name + '\n' + i.address + '\n' + str(i.rssi)))
            self.add_widget(self.label)
            self.add_widget(self.connectbutton)

     def showrgbmatrix(self, client, instance):
        self.disconnect = True 
        print("Client...")
        print(client)        
        self.interesting_device = client     
        self.clear_widgets()
        self.colorpicker = ColorPicker()
        self.add_widget(self.colorpicker)
        self.sendbtn = Button(text = 'Send', size_hint=(0.5,0.1))
        self.add_widget(self.sendbtn)
        self.sendbtn.bind(on_press = partial(self.send_rgb_colors, self.interesting_device))

    async def waiting(self, event):
        await asyncio.sleep(3600)
```

**Reply by ujur007** [reply]
I think I have got some hints. Instead of running the tasks in one single flow, I can put them inside a While() loop. But not sure that ordinary functions(without async) functions will work. Basically, I need them for the button triggers.
I will check a very simple example and let you know.

**Comment by ujur007** [comment]
I think that with Kivy it will be difficult to deal with this issue. Because for buttons it needs only non-synchronous def functions. Which will close the event loop or starting again will give error. 
Another way is to keep the event loop running but then the button's bind methods will not work.!!

**Reply by dlech** [reply]
Are you using `app.async_run("asyncio")` to run the Kivy app? https://kivy.org/doc/master/api-kivy.app.html#asynchronous-app

Anyway, this seems like a discussion for Kivy, not Bleak.

**Reply by ujur007** [reply]
Yes, but there are [issue](https://github.com/kivy/kivy/wiki/Using-Asynchronous-programming-inside-a-Kivy-application) with kivy asyncio. Anyways will try every possible approach.

**Comment by ujur007** [comment]
I can confirm that the problem was not with the Bleak but with Kivy's asyncio implementation. 
Because, I can run the same code with other GUI platforms.
